### PR TITLE
fix findPosition to return only all overlapping positions

### PR DESCRIPTION
**Pull request by ydaveluy**
Add a test to check that TextConsoleViewer::findPosition returns only all overlapping positions.

Fix the findPosition method to be compliant.

The test failed with the old implementation. findPosition must be declared static to verify.

### CODE REVIEW

**Comment by ydaveluy** [comment]
This is related to issue eclipse-platform/eclipse.platform#542

**Comment by SarikaSinha** [comment]
Can you rebase the code after version bump up for tests to go through.

**Comment by ydaveluy** [comment]
Sorry, it is my first contribution and I don't understand what you are expecting :

> Can you rebase the code after version bump up for tests to go through.

**Comment by ydaveluy** [comment]
> Sorry, it is my first contribution and I don't understand what you are expecting :
> 
> > Can you rebase the code after version bump up for tests to go through.

ok got it !

**Comment by SarikaSinha** [comment]
> > Sorry, it is my first contribution and I don't understand what you are expecting :
> > > Can you rebase the code after version bump up for tests to go through.
> 
> ok got it !

Changes look good. 
But you "Did fetch upstream which does merge commit and not rebase".
I understand these steps are not clear.
Please refer to https://github.com/eclipse-pde/eclipse.pde/pull/50#issuecomment-1125062820
As the current branch is not in good shape now, you will have to create new branch and PR.

**Comment by ydaveluy** [comment]
> > > Sorry, it is my first contribution and I don't understand what you are expecting :
> > > > Can you rebase the code after version bump up for tests to go through.
> > 
> > 
> > ok got it !
> 
> Changes look good. But you "Did fetch upstream which does merge commit and not rebase". I understand these steps are not clear. Please refer to [eclipse-pde/eclipse.pde#50 (comment)](https://github.com/eclipse-pde/eclipse.pde/pull/50#issuecomment-1125062820) As the current branch is not in good shape now, you will have to create new branch and PR.

It should be good now.

**Comment by akurtakov** [comment]
I can squash the commits into single one and push it if it's fine with both of you. Committs with message "bugfix" are generally not acceptable as they become meaningless later when one looks at git history.

**Comment by laeubi** [comment]
@ydaveluy do you plan to work on this?

**Comment by ydaveluy** [comment]
Sorry for my late reply, I was in holidays.
@akurtakov yes it is possible to squash, I think that the 4 commit can be squashed together. 
@laeubi according to me the issue eclipse-platform/eclipse.platform#542 is fixed with this PR. An alternative would be to redesign how positions are found, see eclipse-platform/eclipse.platform.debug#46

**Comment by mickaelistria** [comment]
@akurtakov Should we try to have it in RC1, or do you think it's better to delay to 4.26?

**Comment by iloveeclipse** [comment]
I would NOT merge that in RC, too much changes for the last minute fix.

**Comment by mickaelistria** [comment]
> I would NOT merge that in RC, too much changes for the last minute fix.

OK, put 4.26 M1 as milestone then.

**Comment by iloveeclipse** [comment]
> > I would NOT merge that in RC, too much changes for the last minute fix.
> 
> OK, I've put 4.26 M1 as milestone then.

Please retarget in 4.28.

**Comment by mickaelistria** [comment]
@mihnita Can you please review?

**Comment by mihnita** [comment]
I think this PR is really-really outdated, and I am not even sure the bug is a bug.
I've tried to reproduce it, but I was unable to (added a comment there).

And the description of this bug also puzzles me:

> Add a test to check that TextConsoleViewer::findPosition returns only all overlapping positions.
> Fix the findPosition method to be compliant.
> The test failed with the old implementation. findPosition must be declared static to verify.

There is no way that _"the test failed with the old implementation"_ because the test uses reflection to invoke TextConsoleViewer.visitOverlappingPositions, but there is no `visitOverlappingPositions`.

I tried using the existing `Position[] findPosition(int offset, int length, Position[] positions)`, which looks like it was renamed.
But the signature is different, and the method implementation is heavily modified.
So the current test in this commit does not fail for the existing code, as it does not even run with the existing code.

---

TLDR: I will spend some more time trying to understand what the bug is, or if there is a bug to begin with.

**Comment by Shee43** [comment]
Moving this issue to M2 as M1 is done.

File: org.eclipse.debug.tests/src/org/eclipse/debug/tests/console/TextConsoleViewerTest.java [hunk 1]
```diff
@@ -324,8 +324,15 @@ public void testFindPosition() throws Throwable {
 
 			var length = 10;
 			for (int i = 0; i < 100; ++i) {
-				var result = (Position[]) method.invoke(null, i, length, positions);
-				var overlappingPositions = result != null ? Arrays.asList(result) : new ArrayList<>();
+				var overlappingPositions = new ArrayList<>();
+
+				method.invoke(null, i, length, positions, new Consumer<Position>() {
```
Comment (lbownik) [inline org.eclipse.debug.tests/src/org/eclipse/debug/tests/console/TextConsoleViewerTest.java:329]: how about
method.invoke(null, i, length, positions, overlappingPositions::add);

File: org.eclipse.debug.tests/src/org/eclipse/debug/tests/console/TextConsoleViewerTest.java [hunk 2]
```diff
@@ -303,4 +305,49 @@ private static void checkOverlapping(List<StyleRange> styles) {
...
 	}
+
+	/**
+	 * Test that findPosition returns only all overlapping positions
+	 */
+	@Test
+	public void testVisitOverlappingPositions() throws Throwable {
+
+		try {
+			final Method method = TextConsoleViewer.class.getDeclaredMethod("visitOverlappingPositions", int.class, int.class, Position[].class, Consumer.class);
```
Comment (jonahgraham) [inline org.eclipse.debug.tests/src/org/eclipse/debug/tests/console/TextConsoleViewerTest.java:316]: Just wondering if rather than using reflection you can make the method public, but not part of API by using `@noreference`?

I know there are other examples in this test code that do similar, but by using reflection here it means longer term maintenance is harder because things like Refactoring and Call Hierarchy won't show this connection.

**Comment by laeubi** [comment]
This repository was merged with https://github.com/eclipse-platform/eclipse.platform if this PR is still relevant please rebase it and push it to the new origin!
